File: algorithms/sorting/merge_sort/merge_sort.js
function merge(leftArr, rightArr){
    let result = [];
    let leftIndex = 0;
    let rightIndex = 0;
  
    while(leftIndex < leftArr.length && 
          rightIndex < rightArr.length) {
       if(leftArr[leftIndex] < rightArr[rightIndex]) {
         result.push(leftArr[leftIndex]);
         leftIndex++;
       }
       else {
         result.push(rightArr[rightIndex]);
         rightIndex++
      }
    }
  
    return result.concat(leftArr.slice(leftIndex)).concat(rightArr.slice(rightIndex));
  }


function mergeSort (array) {
  if (array.length === 1) {
    return array;
  }

  let length = array.length;
  let middleIndex = Math.floor(length / 2);
  let leftArr = array.slice(0, middleIndex); 
  let rightArr = array.slice(middleIndex);
  
  return merge(
    mergeSort(leftArr),
    mergeSort(rightArr)
  )
}

File: algorithms/sorting/merge_sort/merge_sort.py
def merge(left_arr, right_arr):
    result = []
    left_index, right_index = 0, 0

    while left_index < len(left_arr) and right_index < len(right_arr):
        if left_arr[left_index] < right_arr[right_index]:
            result.append(left_arr[left_index])
            left_index += 1
        else:
            result.append(right_arr[right_index])
            right_index += 1

    return result + left_arr[left_index:] + right_arr[right_index:]


def merge_sort(arr):
    if len(arr) == 1:
        return arr

    middle_index = len(arr) // 2
    left_arr = arr[:middle_index]
    right_arr = arr[middle_index:]

    return merge(
        merge_sort(left_arr),
        merge_sort(right_arr),
    )